File: challenges/2020/07-handyHaversacks/go/challenge/partOne.go
package challenge

func checkBag(ruleset map[string]map[string]int, testCl, targetCl string) bool {
	{
		var isPresent bool
		for v := range ruleset[testCl] {
			if v == targetCl {
				isPresent = true
				break
			}
		}
		if isPresent {
			return true
		}
	}

	for childColour := range ruleset[testCl] {
		if checkBag(ruleset, childColour, targetCl) {
			return true
		}
	}

	return false

}

func PartOne(instr string) int {
	rules := parse(instr)

	canContainTarget := 0

	for bagColour, _ := range rules {
		if bagColour != targetColour {
			if checkBag(rules, bagColour, targetColour) {
				canContainTarget += 1
			}
		}
	}

	return canContainTarget
}

File: challenges/2020/07-handyHaversacks/python/partOne.py
from common import *
from typing import Dict


def check_bag(ruleset: Dict[str, Dict[str, int]], test_cl: str, target_cl: str) -> bool:
    if (
        target_cl in ruleset[test_cl]
    ):  # bag colour can directly contain the target colour
        return True

    for child_colours in ruleset[test_cl]:
        if check_bag(ruleset, child_colours, target_cl):
            return True

    return False


def partOne(instr: str) -> int:
    rules = parse(instr)

    can_contain_target = 0

    for bag_colour in rules:
        if bag_colour != target_colour:
            if check_bag(rules, bag_colour, target_colour):
                can_contain_target += 1

    return can_contain_target
